### PR TITLE
WIP: speed up JumpProblem conversion 2

### PR DESCRIPTION
**Pull request by shashi**
https://gist.github.com/isaacsas/593f9ce7f40adf05b6969ace8fe8a831

with

```julia
const nedgess = 30
const nvertss = 1000
```
gives

```julia
  0.000425 seconds (14.00 k allocations: 641.219 KiB)
Completed: ReactionSystem
  0.033972 seconds (519.34 k allocations: 16.500 MiB)
Completed: convert(JumpSystem)
  0.025654 seconds (585.68 k allocations: 15.495 MiB)
Completed: DiscreteProblem
  2.788977 seconds (2.35 M allocations: 2.143 GiB, 5.16% gc time)
Completed: JumpProblem
  0.000594 seconds (30 allocations: 353.297 KiB)
```

This is probably really bad. I'm thinking of better hash-table operations for the purposes here, hopefully we can turn something from `O(n^2)` to `O(n)`

### CODE REVIEW

**Comment by shashi** [comment]
This is pretty interesting, most of the time is spent in substitute (the call to `subber` in `assemble_maj`), and not the dictionary construction. I guess this should be even faster than what we had before though.

**Comment by ChrisRackauckas** [comment]
What was it before?

**Comment by isaacsas** [comment]
It was pretty bad before, though in the other open PR I had gotten it down a fair amount for everything but the final JumpProblem construction, which was performing badly due to the dependency graph construction among other things.

**Comment by ChrisRackauckas** [comment]
3 seconds still doesn't seem great.

**Comment by shashi** [comment]
Got another 10x

it was doing wasteful dict construction.

```
  0.000401 seconds (14.00 k allocations: 641.219 KiB)
Completed: ReactionSystem
  0.030159 seconds (519.34 k allocations: 16.500 MiB)
Completed: convert(JumpSystem)
  0.024397 seconds (585.68 k allocations: 15.495 MiB)
Completed: DiscreteProblem
  0.311764 seconds (2.18 M allocations: 57.016 MiB)
Completed: JumpProblem
  0.000451 seconds (30 allocations: 353.297 KiB)
Completed: solve
```

**Comment by shashi** [comment]
I think we should only leave `substitue(expr, dict::Dict)` method since everything essentially calls that. This way it's much easier to reason about substitute's performance.

**Comment by shashi** [comment]
Okay another 3x

```
  0.093674 seconds (2.18 M allocations: 57.018 MiB)
```

**Comment by shashi** [comment]
Now with
```julia
const nedgess = 3000
const nvertss = 100000
```
```julia
  0.058344 seconds (1.50 M allocations: 64.072 MiB)
Completed: ReactionSystem
  5.699431 seconds (44.17 M allocations: 1.322 GiB, 44.66% gc time)
Completed: convert(JumpSystem)
  4.851998 seconds (52.66 M allocations: 1.283 GiB, 18.10% gc time)
Completed: DiscreteProblem
 17.310221 seconds (211.11 M allocations: 5.235 GiB, 12.65% gc time)
Completed: JumpProblem
  0.055617 seconds (30 allocations: 34.334 MiB)
Completed: solve
retcode: Default
Interpolation: Piecewise constant interpolation
t: 6-element Array{Float64,1}:
  0.0
  0.1
  0.2
  0.3
  0.4
 20.0
u: 6-element Array{Array{Int64,1},1}:
 [0, 1, 1, 1, 1, 1, 1, 1, 1, 1  …  0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
 [0, 1, 1, 1, 1, 1, 1, 1, 1, 1  …  0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
 [0, 1, 1, 1, 1, 1, 1, 1, 1, 1  …  0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
 [0, 1, 1, 1, 1, 1, 1, 1, 1, 1  …  0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
 [0, 1, 1, 1, 1, 1, 1, 1, 1, 1  …  0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
 [0, 1, 1, 1, 1, 1, 1, 1, 1, 1  …  0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

```

**Comment by shashi** [comment]
```
function states(sys::AbstractSystem)
    unique(isempty(sys.systems) ?
           setdiff(sys.states, value.(sys.pins)) :
           [sys.states;reduce(vcat,namespace_variables.(sys.systems))])
end
```

This is pretty expensive, especially `setdiff`. We should cache this.

Similarly `get_variables!` does `unique` which is expensive too.

**Comment by isaacsas** [comment]
Could also just return `states` if `pins` and `systems` are both empty.

**Comment by shashi** [comment]
> Could also just return states if pins and systems are both empty. I think that’s what I had put in in the other PR.

That's kind of a kludge but ok!

Do you know what's causing the test to fail now?

File: src/systems/jumps/jumpsystem.jl [hunk 1]
```diff
@@ -280,30 +279,30 @@ end
...
     dep
 end
 
-function get_variables!(dep, jump::MassActionJump, variables)
+function get_variables!(dep, jump::MassActionJump, variables::AbstractSet)
     sr = value(jump.scaled_rates)
     (sr isa Symbolic) && get_variables!(dep, sr, variables)
     for varasop in jump.reactant_stoch
-        any(isequal(varasop[1]), variables) && push!(dep, varasop[1])
+        isequal(varasop[1]) in variables && push!(dep, varasop[1])
```
Comment (isaacsas) [inline src/systems/jumps/jumpsystem.jl:291]: Do you really want the `isequal here?

Comment (shashi) [reply src/systems/jumps/jumpsystem.jl:291]: Oh damn my bad!!!

**Comment by isaacsas** [comment]
@shashi Before we merge, can copy the latest timings over here so we have a reference for future comparisons?

**Comment by ChrisRackauckas** [comment]
Was there something done in place of this?

**Comment by isaacsas** [comment]
No, this never got finished and merged unfortunately. If no one else takes a pass I’ll come back to it once I have some time (probably when classes end in a month). There was still more to do.

**Pull request closed without merging**